### ISSUE TITLE
Crash when showing the credits if they're too long

### ISSUE THREAD

**Issue by Konrad22** [wesnoth/wesnoth#3265]
I don't know what is causing it. Reloading and ending again doesn't lead to the crash.

**Comment by sevu** [comment]
I am afraid that this is hard to troubleshoot without more info.
A backtrace or stacktrace would probably help, but ...
 if you are on windows, I don't know how one can get it
 if you are on linux, one can get it with `corecumpctl`

**Comment by Konrad22** [comment]
[wesnoth-20180720-195331-5016.log](https://github.com/wesnoth/wesnoth/files/2215288/wesnoth-20180720-195331-5016.log)
Maybe this can help?

**Comment by gfgtdf** [comment]
ok the logs shows pango erros in a >100000 lines credit text and then gives a 'out of memory'. This makes it seem somewhat likely the out of memeory is related to the lots of text in the credits screen.

**Comment by Konrad22** [comment]
Is the memory, which is avaible for something like this depended on my pc or does it depend on wesnoth?

**Comment by CelticMinstrel** [comment]
Memory available would depend on the computer, not on Wesnoth, but if Wesnoth is allocating a ridiculous amount of memory, then that needs to be fixed somehow.

**Comment by Konrad22** [comment]
How do I check wheter Wesnoth is allocating a ridiculous amount of memory instead of my pc just being bad?

**Comment by gfgtdf** [comment]
> Memory available would depend on the computer, not on Wesnoth

usually this is not the case. Wesnoth releases (at least on windows) is a 32bit exacutable with a maximum of 2gb ram, even if your computer has a lot more ram. On the other hand if the 'real' memeory would go rare then the behviour (at least on windows) would be quite differnt, first the pc would try to swap everying around making everythign extremely slow, then at some point you get a erro message like 'Your computer is low on meemory please close procecces or i will' (not sure on the exact wording here)


In any case this is clearly an bug in wesnoth. I somehow remember that there was at some point a discussion about the screidsscreen allocating one big surface for the scrdits screen, so if we assume each textline to be 20 pixels and the screensize to be 1000 pixels 3000 lines of text and each pixel to be encoded with 4 bytes thats 240,000,000 bytes.

Which is a lot (and clearly too much) but usually not enough to cause an out of memory.

Which campaign did you play when you get this error? (and with mods or not)

**Comment by Konrad22** [comment]
So, I apparently overlooked that I was asked a question. I do not remember what campaign I played back then, but this issue is still happening to me and it's happening whenever I finish any campaign. I never use mods in SP.
List of (somewhat) finished campaigns since back then:
Wings of Victory
Bad Moon Rising 2
TSGrewrite
After the Storm 2
Panther Lord
TSG
SoF
DW
NR
SotA

**Comment by Konrad22** [comment]
It happens when I start wesnoth, load the epilogue, skip the text. The 'The End' is written on my screen and before the credits appear, the game crashes.

The one from yesterday:
[wesnoth-20181214-232657-3124.log](https://github.com/wesnoth/wesnoth/files/2682597/wesnoth-20181214-232657-3124.log)

The one just now with only loading the epilogue:
[wesnoth-20181215-131244-6880.log](https://github.com/wesnoth/wesnoth/files/2682598/wesnoth-20181215-131244-6880.log)

They both end with 'Text is too long to render'.

**Comment by jostephd** [comment]
That's from here:

https://github.com/wesnoth/wesnoth/blob/571da300784253ecc9a77ba215acb037ecee04fd/src/font/text.cpp#L646-L647

**Comment by gfgtdf** [comment]
Maybe it'd be better to just show the credits of the campaign played (instead of all campaigns installed) when finishing a game.

**Comment by CelticMinstrel** [comment]
It does seem silly to show the credits from campaigns you weren't even playing, yes.

**Comment by CelticMinstrel** [comment]
(That said, if you click "credits" at the main menu, it's not so silly...)

**Comment by jostephd** [comment]
I can reproduce this.

I edited data/core/about.cfg and changed `"josteph" ` to `"josteph\n..."` where `...` is the [your log file](https://github.com/wesnoth/wesnoth/issues/3265#issuecomment-447564511) filtered through `sed s/././g` (size: 1751 lines, 38716 bytes) and then opened the credits screen in a wesnoth environment that has a large number of add-ons installed.

I got the same error you did:

```
' has unescaped ampersands '&', escaped them.
Caught general 'St12length_error' exception:
Text is too long to render
```

**Comment by Konrad22** [comment]
Is there any chance that this gets added as a 1.14.13 milestone?

**Comment by Konrad22** [comment]
FTR, just now I tried to reproduce it on 1.15.3, and I could reproduce it indeed.

**Comment by Konrad22** [comment]
That's the 'new' error message, but it's the same result.
```' has unescaped ampersands '&', escaped them.
20200621 21:47:09 error gui/layout: pango_text::set_markup text '<span size='xx-large'>Bad Moon Rising' has broken markup, set to normal text.
Caught general 'St12length_error' exception:
basic_string::_M_create```

**Comment by stevecotton** [comment]
> That's the 'new' error message, but it's the same result.

Confirmed - my previous change increased the amount of text to make it crash, but I now realise that it still crashes and that the new failure is more confusing to debug; sorry.There are a couple of different ways that the current implementation can crash because of the huge single surface:
* In src/font/text.cpp, the line `this->create_surface_buffer(stride * height);` is a 32-bit multiplication, even on some 64-bit architectures. Wrap-round here results in a segfault during the `render(*part_layout, part_rect, cumulative_height * stride, stride);`
* Later on, an SDL `surface` might have a size of 0x0 pixels, which I assume is caused by SDL having a sensible sanity check on the size. This triggers an assert that the size is greater than zero.

Does anyone want to redesign the UI before we look at how to implement a better scrolling box? My suggestion:
* the end-of-campaign credits could show just the campaign's and era's credits
* the main-menu's credits button could use the same GUI as the help system, with a separate page for each add-on

**Comment by Pentarctagon** [comment]
> the end-of-campaign credits could show just the campaign's and era's credits

This sounds like a good thing to do regardless of this bug, IMO.

> the main-menu's credits button could use the same GUI as the help system, with a separate page for each add-on

The ingame Help still uses GUI1(#3653), so that would be a backward step in that sense.

**Comment by stevecotton** [comment]
I think #4960 is unrelated to this issue, it seems that is in the outro and not the credits. However, having looked again at this issue I think reopening it was a good thing, as I'm unhappy with my original fix.

**Comment by stevecotton** [comment]
I've found a new testcase - after installing the current 1.14 version of After The Storm on 1.15, those credits crash 1.15 (but not 1.14).

**Comment by stevecotton** [comment]
I've just caught up with the IRC / Discord discussion on October 6th. If you try to reimplement the credits using GUI2's listboxes, then it still tries to create a background for a huge widget, and runs out of memory at that point. AFAICS the GUI2 drawing routines support drawing without starting at the top-left, but don't have a way to clip the bottom-right part of the widget.

There's some prototyping of this in 507a93bae9a07c770b28e90ce4943851d333b8b0 in my prototype/listbox_credits_example branch. The stack-trace as it fails starts:
1. surface::surface(int, int) (w=1528, h=745360)
2. gui2::canvas::draw(bool) (force=false)
3. gui2::canvas::blit(surface&, SDL_Rect) (surf=..., rect=...)
4. gui2::styled_widget::impl_draw_background(surface&, int, int) (frame_buffer=..., x_offset=0, y_offset=0)
5. gui2::toggle_panel::impl_draw_background(surface&, int, int) (frame_buffer=..., x_offset=0, y_offset=0)
6. gui2::widget::draw_background(surface&, int, int) (frame_buffer=..., x_offset=0, y_offset=0)
7. gui2::grid::impl_draw_children(surface&, int, int) (frame_buffer=..., x_offset=0, y_offset=0)
8. gui2::widget::draw_children(surface&, int, int) (frame_buffer=..., x_offset=0, y_offset=0)

**Comment by stevecotton** [comment]
Fixed in #5682.
